### PR TITLE
Feature/spy call order by parameter

### PR DESCRIPTION
**Pull request by cnexans**
#59 

Adds the following assertions

```
to.have.been.first.called.with
to.have.been.second.called.with
to.have.been.third.called.with
on.nth(i).be.called.with
```

Regards

### CODE REVIEW

**Comment by stalniy** [comment]
Good job!

**Comment by cnexans** [comment]
Thank you @stalniy! Looking forward to contribute.

Regards

**Comment by keithamus** [comment]
This is looking good so far. I'd also like to see some tests that address partial arguments - for example what do these do?

```js
spy(1,2,3)
spy(3,4,5)
// These should definitely pass - let's see tests for that
expect(spy).first.to.be.called.with(1,2,3)
expect(spy).second.to.be.called.with(3,4,5)
expect(spy).first.to.be.called.exactly.with(1,2,3)
expect(spy).second.to.be.called.exactly.with(3,4,5)

// These should definitely fail - let's see tests for that
expect(spy).first.to.be.called.with(3,4,5)
expect(spy).second.to.be.called.with(1,2,3)
expect(spy).first.to.be.called.exactly.with(1)
expect(spy).first.to.be.called.exactly.with(2)
expect(spy).first.to.be.called.exactly.with(3)
expect(spy).second.to.be.called.exactly.with(3)
expect(spy).second.to.be.called.exactly.with(4)
expect(spy).second.to.be.called.exactly.with(5)


// These should definitely pass - let's see tests for that
expect(spy).first.to.not.be.called.with(3,4,5)
expect(spy).second.to.not.be.called.with(1,2,3)
expect(spy).first.to.not.be.called.exactly.with(1)
expect(spy).first.to.not.be.called.exactly.with(2)
expect(spy).first.to.not.be.called.exactly.with(3)
expect(spy).second.to.not.be.called.exactly.with(3)
expect(spy).second.to.not.be.called.exactly.with(4)
expect(spy).second.to.not.be.called.exactly.with(5)

// Should these all pass? My intuition says yes
expect(spy).first.to.be.called.with(2)
expect(spy).first.to.be.called.with(2,3)
expect(spy).first.to.be.called.with(1)
expect(spy).first.to.be.called.with(1,2)
expect(spy).second.to.be.called.with(3,4)
expect(spy).second.to.be.called.with(4,5)
expect(spy).first.to.be.called.with(3)
expect(spy).second.to.be.called.with(3)

// Should these all fail? My intuition says yes
expect(spy).first.to.not.be.called.with(2)
expect(spy).first.to.not.be.called.with(2,3)
expect(spy).first.to.not.be.called.with(1)
expect(spy).first.to.not.be.called.with(1,2)
expect(spy).second.to.not.be.called.with(3,4)
expect(spy).second.to.not.be.called.with(4,5)
expect(spy).first.to.not.be.called.with(3)
expect(spy).second.to.not.be.called.with(3)
```

**Comment by cnexans** [comment]
Hello @keithamus  I would like to have a good definition of nth with and nth exactly with. This is what I'm thinking and please correct me if you don't agree:

1. nth with ~ given (spy, n, expectedArguments) => the nth call of the spy should be a `subset` of expectedArguments
2. nth exactly with ~ given (spy, n, expectedArguments) => the nth call of the spy should be `equal as array` to expectedArguments

At the moment I am receiving `Error: Invalid Chai property: exactly.with. See docs for proper usage of "exactly".`

The first part is actually covered as all the tests without exactly pass or fail as expected. I will update the PR to fit spec 2, this should only require:

* Add exactly as a property flag
* Add exactly case to with method

Regards

**Comment by keithamus** [comment]
Sorry yeah, I forgot the syntax - it should be `with.exactly` not `exactly.with` I'll fix my comment. So `.called.with` - as you rightly say - only ever asserts for _subsets_ of arguments, while `.called.with.exactly` should test the whole argument set. I believe the same should apply to the `nth` flags.

**Comment by cnexans** [comment]
@keithamus Sorry for the delay, I had a complicated week. I just added first, second, third and nth with.exactly assertions.

This is the thing I found about your last post:

```
spy(1, 2, 3)

// passes as {1} is a subset of {1, 2, 3} --- {1, 2, 3} is a superset of {1}
spy.should.be.first.called.with(1)

// passes as {3, 2, 1} is actually a subset of {1, 2, 3}, order does not matter!
spy.should.be.first.called.with(3, 2, 1)

// fails as {4} is not a subset of {1, 2, 3}
spy.should.be.first.called.with(4)

// fails as {1, 2, 4} is not a subset of {1, 2, 3}
spy.should.be.first.called.with(4)

// fails as for each element of {{1, 2, 2}} is not true that its multiplicity is equal to its multiplicity in {{1, 2, 3}}, denoting {{...}} as multiset
spy.should.be.first.called.with(1, 2, 2)

// passes as [1, 2, 3] is equal to [1, 2, 3]
spy.should.be.first.called.with.exactly(1, 2, 3)

// fails as [2, 1, 3] is not equal to [1, 2, 3], order matters!
spy.should.be.first.called.with.exactly(2, 1, 3)

// passes as [1, 2] is not equal to [1, 2, 3], length matters!
spy.should.be.first.called.with.exactly(1, 2)
```

This open my mind to a `some` assertion, I may open an issue on this matter.

File: test/spies.js [hunk 1]
```diff
@@ -357,47 +434,65 @@ describe('Chai Spies', function () {
     });
   });
 
-  describe('.always.with.exactly(arg, ...)', function () {
-    it('should pass when called with an argument', function () {
+  describe('.nth(...).with.exactly(arg, ...)', function () {
```
Comment (keithamus) [inline test/spies.js]: Curious here why `always.with.exactly` tests have been removed?

Comment (cnexans) [inline test/spies.js]: Hey, seems I accidentally replaced those tests. I will be restoring them.

**Comment by cnexans** [comment]
Tests appearently lost, restored. I apologize for that mistake. :)

**Comment by cnexans** [comment]
Hello, are there any news on this?

File: README.md [hunk 2]
```diff
@@ -178,6 +178,37 @@ expect(spy).to.have.been.called.always.with.exactly('foo');
...
+spy.should.on.nth(5).be.called.with('foobar');
+```
+
+These assertions requires the spy to be called at least the 
+number of times required, for example
+
+```js
+spy('foo');
+spy('bar');
+expect(spy).to.have.been.first.called.with('baz');
```
Comment (stalniy) [inline README.md]: I guess it should be `to.have.been.THIRD.called` as you then say that spy has not been called third time.

File: lib/spy.js [hunk 3]
```diff
@@ -433,41 +433,111 @@ module.exports = function (chai, _) {
...
         , expArgs
       );
+    } else if (nthCall) {
+      var passed = spy.calls[nthCall],
+          ordinalNumber = generateOrdinalNumber(nthCall),
+          actArgs = calls[nthCall];
+      new Assertion(this._obj).to.be.have.been.called.min(nthCall);
+      this.assert(
+          nthCallWith(spy, nthCall - 1, expArgs)
+        , 'expected ' + this._obj + ' to have been called at the ' + ordinalNumber + ' time with #{exp} but got ' + actArgs
```
Comment (stalniy) [inline lib/spy.js]: `actArgs` is an array, if so then better to use `#{act}` and pass 5th argument in `this.assert`

File: lib/spy.js [hunk 4]
```diff
@@ -521,6 +592,16 @@ module.exports = function (chai, _) {
         , 'expected ' + this._obj + ' to have not always been called with exactly #{exp}'
         , args
       );
+    } else if(_with && nthCall) {
+      var ordinalNumber = generateOrdinalNumber(nthCall),
+          actArgs = calls[nthCall - 1];
+      new Assertion(this._obj).to.be.have.been.called.min(nthCall);
+      this.assert(
+          _.eql(actArgs, args)
+        , 'expected ' + this._obj + ' to have been called at the ' + ordinalNumber + ' time with exactly #{exp} but got ' + args
```
Comment (stalniy) [inline lib/spy.js]: same here I guess `args` can be an array of objects and if so, then you will have not well formatted error messages

**Comment by stalniy** [comment]
@cnexans do you have time to fix this?

**Comment by cnexans** [comment]
Hi @stalniy I'm sorry I've been busy these days. I'll be fixing this next days and report back to you

**Comment by stalniy** [comment]
@cnexans any progress?

**Comment by stalniy** [comment]
@cnexans @keithamus I fixed the issues and squashed commits into one. So, now it looks good and can be merged. By the way found a bug with imcorrect arguments call comparison. Fixed that as well :)

**Pull request merged**